File: skills/coordinator/watchchan.py
"""Reader/drainer half of the watcher->coord message channel (Slice 1/4).

The Go writer (internal/watchchan) appends one file per EVENT message under
~/.fleet/projects/<project>/watch-msgs/ and overwrites one file per WORKER
heartbeat (hb-<worker>.json). This module is the coord-side reader: it lists
pending EVENTS in time order, reads/refuses-newer a message, and provides
mark_done (the apply-before-delete commit point). Heartbeats are read
separately and are NEVER drained or deleted as events.

Mirrors the internal/queue list+delete idiom and pr_watch.py's
refuse-newer-schema discipline. No coordinator.lock is taken here — this is
pre-serialization plumbing.

    EVENTS    <unix_nanos>-<watcher_id>-<seq>.json   list_pending / mark_done
    HEARTBEAT hb-<worker_id>.json                    read_heartbeat (latest)

Drain contract (apply-before-delete):

    for path in list_pending(project_dir):
        msg = read_message(path)     # refuses a newer schema v
        apply(msg)                   # <-- COMMIT POINT is AFTER this
        mark_done(path)              # a crash before here replays the msg

The channel does NOT dedup; replay-idempotency is the caller's job
(bookkeeping flips are idempotent; dispatch suppression uses pr-watch keys).
"""
from __future__ import annotations

import json
import os
from pathlib import Path
from typing import Optional

# Wire version shared with internal/watchchan.SchemaVersion. We refuse a
# message whose `v` is newer than this rather than mis-parse it.
SCHEMA_VERSION = 1

DIR_NAME = "watch-msgs"
HEARTBEAT_PREFIX = "hb-"


class SchemaTooNewError(Exception):
    """A message declares a schema `v` newer than this reader understands."""


def watch_msgs_dir(project_dir) -> Path:
    """Return ~/.fleet/projects/<project>/watch-msgs/ for a project dir."""
    return Path(project_dir) / DIR_NAME


def _is_event_file(name: str) -> bool:
    """True iff name is a drainable EVENT message — a .json file that is
    neither a heartbeat (hb-*) nor a WriteAtomic .tmp.<pid> sidecar."""
    if name.startswith(HEARTBEAT_PREFIX):
        return False
    # WriteAtomic sidecars are "<name>.tmp.<pid>" — they do not end in .json.
    return name.endswith(".json")


def list_pending(project_dir) -> list[Path]:
    """Return absolute paths of pending EVENT messages, sorted by filename
    (== time order, since the filename is <unix_nanos>-...). Excludes hb-*
    heartbeats AND .tmp sidecars; a half-written file is never returned. A
    missing directory yields an empty list (not an error)."""
    d = watch_msgs_dir(project_dir)
    try:
        names = os.listdir(d)
    except FileNotFoundError:
        return []
    events = [n for n in names if _is_event_file(n)]
    events.sort()
    return [d / n for n in events]


def read_message(path) -> dict:
    """Parse one message file. Refuses a `v` newer than SCHEMA_VERSION
    (SchemaTooNewError) rather than mis-parsing — mirrors
    queue.ReadSpawnFresh / pr_watch SCHEMA_VERSION discipline."""
    with open(path, "r", encoding="utf-8") as fh:
        msg = json.load(fh)
    if not isinstance(msg, dict):
        raise ValueError(f"watchchan: {path} is not a JSON object")
    v = msg.get("v")
    if isinstance(v, int) and v > SCHEMA_VERSION:
        raise SchemaTooNewError(
            f"watchchan: {path} schema v={v} newer than supported {SCHEMA_VERSION} (upgrade fleet)"
        )
    return msg


def _fsync_dir(directory) -> None:
    """Best-effort fsync of a directory so a just-completed unlink is durable
    across a crash. On POSIX an unlink is NOT crash-durable until its parent
    directory is fsynced; without this a "drained" event file can reappear
    after a reboot and be processed twice. Mirrors the Go writer's fsyncDir.
    Errors are swallowed: the unlink already took observable effect, and some
    platforms/filesystems refuse a directory fsync."""
    try:
        fd = os.open(str(directory), os.O_RDONLY)
    except OSError:
        return
    try:
        os.fsync(fd)
    except OSError:
        pass
    finally:
        os.close(fd)


def mark_done(path) -> None:
    """Delete a drained EVENT file — the apply-before-delete COMMIT POINT.
    The caller applies the mutation first, THEN calls mark_done; a crash in
    between replays the message on recovery. Missing file is a no-op
    (idempotent).

    The unlink is made durable (parent-dir fsync) so this is a REAL commit
    point: without it, applying a worker_delta/pr_event, calling mark_done,
    then crashing could resurrect the file on reboot and double-apply the
    action. A missing file skips the fsync (nothing was committed here)."""
    path = Path(path)
    try:
        os.remove(path)
    except FileNotFoundError:
        return
    _fsync_dir(path.parent)


# --- heartbeats: latest-state, read separately, never drained as events ---


def heartbeat_path(project_dir, worker_id: str) -> Path:
    """Return the hb-<worker>.json path for a worker."""
    return watch_msgs_dir(project_dir) / f"{HEARTBEAT_PREFIX}{worker_id}.json"


def read_heartbeat(project_dir, worker_id: str) -> Optional[dict]:
    """Return the latest heartbeat dict for a worker, or None if none/
    unreadable. Heartbeats are latest-state files — read, never mark_done.

    Refuses a `v` newer than SCHEMA_VERSION (SchemaTooNewError), same as
    read_message: in a mixed-version rollout an older coord must surface the
    upgrade requirement, NOT silently consume an incompatible heartbeat and
    skew the liveness/stuck-worker decisions built on it. Raising (rather than
    returning None) is deliberate — a None would read as 'worker dead' and
    could trip false stuck-worker recovery."""
    p = heartbeat_path(project_dir, worker_id)
    try:
        with open(p, "r", encoding="utf-8") as fh:
            hb = json.load(fh)
    except (FileNotFoundError, json.JSONDecodeError):
        return None
    if isinstance(hb, dict):
        v = hb.get("v")
        if isinstance(v, int) and v > SCHEMA_VERSION:
            raise SchemaTooNewError(
                f"watchchan: {p} heartbeat schema v={v} newer than supported "
                f"{SCHEMA_VERSION} (upgrade fleet)"
            )
    return hb


def list_heartbeats(project_dir) -> list[Path]:
    """Return absolute paths of every hb-<worker>.json heartbeat file."""
    d = watch_msgs_dir(project_dir)
    try:
        names = os.listdir(d)
    except FileNotFoundError:
        return []
    return [
        d / n
        for n in names
        if n.startswith(HEARTBEAT_PREFIX) and n.endswith(".json")
    ]

File: internal/watchchan/watchchan.go
// Package watchchan is the append-safe, crash-safe message channel that
// lets many read-only *watchers* report events to the single *coord* (the
// dispatch controller) without racing. It is the foundation slice of
// DESIGN-coord-supervisor-in-daemon: the substrate Slices 2-4 build on.
//
// Why not internal/queue? The queue writes one file per AGENT and
// re-enqueuing CLOBBERS the prior message (state.WriteAtomic overwrites a
// stable name). That is correct for its per-agent spawn-fresh use, but it
// cannot fan IN N watchers — a second watcher's report would erase the
// first. This channel instead writes one file per MESSAGE under a
// per-project directory, so N producers append concurrently and never
// clobber each other.
//
// Two message disciplines, because events and heartbeats have opposite
// needs:
//
//	EVENTS (worker_delta, pr_event)   — unique-name append, never clobbered.
//	  filename: <unix_nanos>-<watcher_id>-<seq>.json
//	  uniqueness triple: (unix_nanos, watcher_id, seq)
//	    - watcher_id embeds the pid, so two PROCESSES never share one
//	      (a per-process seq alone can't stop a cross-process clobber).
//	    - seq is a per-process atomic counter breaking same-nanos ties
//	      WITHIN a process.
//	  drained by the Python reader (watchchan.py) apply-before-delete.
//
//	HEARTBEATS (liveness_heartbeat) — keyed by WORKER, overwrite-in-place.
//	  filename: hb-<worker_id>.json  (NO pid — worker-keyed)
//	  A latest-state file, atomically overwritten each probe. Because the
//	  name depends only on the worker, a watcher restart (new pid)
//	  overwrites the SAME file in place rather than orphaning a stale one,
//	  so heartbeats are structurally bounded at <= N_workers regardless of
//	  watcher churn. They are NEVER drained/deleted as events and are
//	  CAP-EXEMPT (dropping a heartbeat under cap pressure would skew the
//	  stuck-timer Slice 3 derives from it).
//
// Both writers reuse state.WriteAtomic (tmp+fsync+rename+parent-fsync); no
// lock is taken in this layer — filename uniqueness (events) and stable-name
// last-rename-wins overwrite (heartbeats) make them multi-producer-safe.
package watchchan

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"strconv"
	"strings"
	"sync/atomic"
	"time"

	"github.com/edisonshen/fleet/internal/state"
)

// SchemaVersion is the wire version shared by the Go writer and the Python
// reader (skills/coordinator/watchchan.py SCHEMA_VERSION). The reader
// refuses a message whose `v` is newer than it understands rather than
// mis-parsing — mirroring internal/queue.ReadSpawnFresh's refuse-newer
// discipline.
const SchemaVersion = 1

// DirName is the per-project subdirectory holding the channel:
// ~/.fleet/projects/<project>/watch-msgs/.
const DirName = "watch-msgs"

// HeartbeatPrefix names the worker-keyed latest-state heartbeat files. The
// event drain skips any file with this prefix; the cap never counts or
// drops them.
const HeartbeatPrefix = "hb-"

// Message kinds.
const (
	KindWorkerDelta = "worker_delta"       // an actionable worker-liveness change
	KindPREvent     = "pr_event"           // a PR/CI state move (carries SHAs)
	KindHeartbeat   = "liveness_heartbeat" // a per-epoch liveness probe (heartbeat file)
)

// DefaultCap bounds the number of EVENT messages on disk. When exceeded the
// writer drops the oldest events (logging each) so an undrained channel
// (coord dead) cannot grow without bound. Heartbeats are NOT in this cap.
const DefaultCap = 10000

// Message is the self-describing wire shape, identical on both sides.
//
// `seq` is intentionally NOT a field here: it is a filename-only
// disambiguator for events, not part of the delivered payload.
type Message struct {
	V         int            `json:"v"`          // schema version (refuse-newer on read)
	Kind      string         `json:"kind"`       // one of the Kind* constants
	WatcherID string         `json:"watcher_id"` // process-unique (embeds pid) — see WatcherID
	EpochID   int64          `json:"epoch_id"`   // watcher's per-pass probe counter (NOT a dispatch_generation)
	Key       string         `json:"key"`        // delivery key; PR events include the head SHA
	Payload   map[string]any `json:"payload,omitempty"`
	TS        string         `json:"ts"` // UTC-ISO human/audit field (filename nanos is the ordering key)
}

// nowNanos is the clock seam (overridable in tests to force same-nanosecond
// appends). Production reads the monotonic-free wall clock in nanoseconds.
var nowNanos = func() int64 { return time.Now().UnixNano() }

// seqCounter is the per-process monotonic sequence. atomic so concurrent
// AppendEvent calls in one process each get a distinct value, breaking
// same-nanosecond ties. Combined with the pid-bearing watcher_id this makes
// the filename unique across ALL producers.
var seqCounter atomic.Uint64

// WatcherID returns a process-unique watcher id by appending this process's
// pid to base (e.g. "<agent_id>" -> "<agent_id>-<pid>"). The pid is
// load-bearing: it is what keeps two watcher PROCESSES from emitting the
// same event filename at the same nanosecond+seq. It is the EVENT-filename
// component only; heartbeats are keyed by worker, so a watcher restart never
// churns heartbeat files.
func WatcherID(base string) string {
	return base + "-" + strconv.Itoa(os.Getpid())
}

// Channel is the per-project message channel rooted at
// ~/.fleet/projects/<project>/watch-msgs/.
type Channel struct {
	dir  string                           // the watch-msgs directory
	cap  int                              // event-message cap (0 = unbounded)
	logf func(format string, args ...any) // drop-logging sink (never silent)
}

// New builds a Channel under projectDir (typically the result of
// state.ProjectDir(project)). The watch-msgs directory is created lazily on
// first write.
func New(projectDir string) *Channel {
	return &Channel{
		dir:  filepath.Join(projectDir, DirName),
		cap:  DefaultCap,
		logf: func(format string, args ...any) { fmt.Fprintf(os.Stderr, "watchchan: "+format+"\n", args...) },
	}
}

// Dir returns the absolute watch-msgs directory path.
func (c *Channel) Dir() string { return c.dir }

// eventFilename builds the unique event filename. unix_nanos is zero-padded
// to 19 digits so lexicographic sort == time order (int64 nanos stays 19
// digits until year 2262), and the (unix_nanos, watcher_id, seq) triple is
// what guarantees no clobber across producers.
func eventFilename(unixNanos int64, watcherID string, seq uint64) string {
	return fmt.Sprintf("%019d-%s-%d.json", unixNanos, watcherID, seq)
}

// AppendEvent publishes msg as a new unique-name event file and returns its
// absolute path. msg.WatcherID must be process-unique (use WatcherID). The
// channel fills v/ts and assigns the filename nanos+seq. After writing it
// enforces the event cap (heartbeats unaffected).
func (c *Channel) AppendEvent(msg Message) (string, error) {
	if msg.WatcherID == "" {
		return "", fmt.Errorf("watchchan: AppendEvent requires WatcherID")
	}
	if msg.Kind == KindHeartbeat {
		return "", fmt.Errorf("watchchan: heartbeats use WriteHeartbeat, not AppendEvent")
	}
	if msg.Kind == "" {
		return "", fmt.Errorf("watchchan: AppendEvent requires Kind")
	}
	stamp(&msg)
	if err := os.MkdirAll(c.dir, 0o755); err != nil {
		return "", fmt.Errorf("watchchan: mkdir %s: %w", c.dir, err)
	}
	data, err := marshal(msg)
	if err != nil {
		return "", err
	}
	seq := seqCounter.Add(1)
	path := filepath.Join(c.dir, eventFilename(nowNanos(), msg.WatcherID, seq))
	if err := state.WriteAtomic(path, data); err != nil {
		return "", err
	}
	c.enforceCap()
	return path, nil
}

// WriteHeartbeat publishes a worker-keyed latest-state heartbeat at
// hb-<workerID>.json, atomically overwriting any prior heartbeat for that
// worker. Heartbeats are never drained as events and never counted against
// the cap.
//
// Concurrency model: safe across PROCESSES (the real case — old+standby
// watchers overlapping during a handoff). WriteAtomic's temp file is
// pid-scoped, so two processes use distinct temps and the atomic rename
// gives last-rename-wins, which is benign for a liveness signal (either
// proves the worker live; the latest epoch wins). It is NOT a guarantee for
// two goroutines in ONE process writing the SAME worker's heartbeat
// concurrently — they would share the pid-scoped temp. That never happens by
// construction: a watcher probes each worker sequentially within a pass, and
// distinct workers have distinct filenames.
func (c *Channel) WriteHeartbeat(workerID string, msg Message) (string, error) {
	if workerID == "" {
		return "", fmt.Errorf("watchchan: WriteHeartbeat requires workerID")
	}
	msg.Kind = KindHeartbeat
	stamp(&msg)
	if err := os.MkdirAll(c.dir, 0o755); err != nil {
		return "", fmt.Errorf("watchchan: mkdir %s: %w", c.dir, err)
	}
	data, err := marshal(msg)
	if err != nil {
		return "", err
	}
	path := filepath.Join(c.dir, heartbeatName(workerID))
	if err := state.WriteAtomic(path, data); err != nil {
		return "", err
	}
	return path, nil
}

// RemoveHeartbeat reaps the heartbeat for a worker that has terminated. The
// live watcher calls this on observing the worker gone (no watcher-restart
// orphan exists — the stable name overwrites in place). Missing file is a
// no-op.
func (c *Channel) RemoveHeartbeat(workerID string) error {
	path := filepath.Join(c.dir, heartbeatName(workerID))
	if err := os.Remove(path); err != nil && !os.IsNotExist(err) {
		return fmt.Errorf("watchchan: remove heartbeat %s: %w", path, err)
	}
	return nil
}

// ReapOrphanHeartbeats is the gc backstop: it removes any hb-<worker>.json
// whose worker is not in live. Keyed on worker-existence — the only reap
// trigger needed, since the stable name means no watcher-restart orphan can
// occur. Returns the paths reaped.
func (c *Channel) ReapOrphanHeartbeats(live map[string]bool) ([]string, error) {
	entries, err := os.ReadDir(c.dir)
	if err != nil {
		if os.IsNotExist(err) {
			return nil, nil
		}
		return nil, fmt.Errorf("watchchan: readdir %s: %w", c.dir, err)
	}
	var reaped []string
	for _, e := range entries {
		if e.IsDir() {
			continue
		}
		worker, ok := workerFromHeartbeat(e.Name())
		if !ok {
			continue
		}
		if live[worker] {
			continue
		}
		p := filepath.Join(c.dir, e.Name())
		if err := os.Remove(p); err != nil && !os.IsNotExist(err) {
			return reaped, fmt.Errorf("watchchan: reap %s: %w", p, err)
		}
		reaped = append(reaped, p)
	}
	return reaped, nil
}

// enforceCap drops the oldest EVENT messages once the event count exceeds
// c.cap, logging each drop (never silent). Heartbeats and .tmp sidecars are
// not counted or dropped.
func (c *Channel) enforceCap() {
	if c.cap <= 0 {
		return
	}
	events := c.listEventNames()
	if len(events) <= c.cap {
		return
	}
	// names sort lexicographically == oldest-first (nanos-prefixed).
	drop := events[:len(events)-c.cap]
	for _, name := range drop {
		p := filepath.Join(c.dir, name)
		if err := os.Remove(p); err != nil && !os.IsNotExist(err) {
			c.logf("cap %d exceeded but failed to drop %s: %v", c.cap, name, err)
			continue
		}
		c.logf("cap %d exceeded, dropped oldest event %s", c.cap, name)
	}
}

// listEventNames returns event filenames (sorted oldest-first), excluding
// heartbeats and .tmp sidecars.
func (c *Channel) listEventNames() []string {
	entries, err := os.ReadDir(c.dir)
	if err != nil {
		return nil
	}
	var out []string
	for _, e := range entries {
		if e.IsDir() {
			continue
		}
		if isEventFile(e.Name()) {
			out = append(out, e.Name())
		}
	}
	sort.Strings(out)
	return out
}

// isEventFile reports whether name is a drainable event message (a .json
// file that is not a heartbeat and not a .tmp sidecar).
func isEventFile(name string) bool {
	if strings.HasPrefix(name, HeartbeatPrefix) {
		return false
	}
	if filepath.Ext(name) != ".json" {
		return false // .tmp.<pid> sidecars end in the pid, not .json
	}
	return true
}

func heartbeatName(workerID string) string { return HeartbeatPrefix + workerID + ".json" }

// workerFromHeartbeat extracts the worker id from hb-<worker>.json, or
// (\"\", false) if name is not a heartbeat file.
func workerFromHeartbeat(name string) (string, bool) {
	if !strings.HasPrefix(name, HeartbeatPrefix) || !strings.HasSuffix(name, ".json") {
		return "", false
	}
	return strings.TrimSuffix(strings.TrimPrefix(name, HeartbeatPrefix), ".json"), true
}

// stamp fills the wire fields the channel owns: the schema version (always
// the current one — the writer never emits an old shape) and a default
// UTC-ISO ts when the caller left it empty.
func stamp(msg *Message) {
	msg.V = SchemaVersion
	if msg.TS == "" {
		msg.TS = time.Now().UTC().Format(time.RFC3339)
	}
}

func marshal(msg Message) ([]byte, error) {
	data, err := json.MarshalIndent(msg, "", "  ")
	if err != nil {
		return nil, fmt.Errorf("watchchan: marshal: %w", err)
	}
	return append(data, '\n'), nil
}
